### PR TITLE
[CHORE] Refactor mention tracking logic

### PR DESCRIPTION
**Pull request by sumukhah**
This PR focuses on improving command, emoji, channel and user mentions and also Removed some unnecessary codes.

## Proposed changes

If a word starts with symbols(@, #, :, /) then only suggestion Pop ups.

## Issue(s)	

Fixes: #2996 

## How to test or reproduce

Use the symbols in the middle of sentences, Pop-ups won't happen. If we use it at the beginning, it will pop up.
## Screenshots
**Using symbols at the beginning**
Screenshot 2021-03-18 at 11 02 34 PM

**Using symbols in the middle**
Screenshot 2021-03-18 at 11 02 20 PM

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves a current function)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [x] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by diegolmello** [comment]
@sumukhah It looks awesome! Can you write e2e tests? Thanks for your contribution!

**Review by kresnaputra — APPROVED** [review]
LGTM

**Comment by sumukhah** [comment]
Please review @diegolmello

File: app/containers/MessageBox/index.js [hunk 1]
```diff
@@ -352,60 +353,33 @@ class MessageBox extends Component {
...
-				const [, lastChar, name] = result;
-				this.identifyMentionKeyword(name, lastChar);
-			} catch (e) {
-				log(e);
-			}
+		const txt = text.split(' ');
+		const lastWord = txt[txt.length - 1];
+		const result = lastWord.substring(1);
+
+		const commandMention = lastWord.match(/^\//);
```
Comment (diegolmello) [inline app/containers/MessageBox/index.js:368]: Slash commands should run only at the beginning of the message.

This should be considered/listened as slash commands:
```
/giphy
```

This should not:
```
Hey, check this: /giphy
```
In the message above, slash commands shouldn't be triggered when the user typed `/`

File: app/containers/MessageBox/index.js [hunk 2]
```diff
@@ -352,60 +353,33 @@ class MessageBox extends Component {
...
-				let regexp = /(#|@|:|^\/)([a-z0-9._-]+)$/im;
-
-				// if sharing, track #|@|:
-				if (sharing) {
-					regexp = /(#|@|:)([a-z0-9._-]+)$/im;
-				}
-
-				const result = lastNativeText.substr(0, cursor).match(regexp);
-				if (!result) {
-					if (!sharing) {
```
Comment (diegolmello) [inline app/containers/MessageBox/index.js:394]: As wrote on this condition, slash commands shouldn't be triggered when `sharing === true`.
You can reproduce it by sharing images and then starting a message with `/`.

File: app/containers/MessageBox/index.js [hunk 3]
```diff
@@ -352,60 +353,33 @@ class MessageBox extends Component {
...
-		const db = database.active;
+	debouncedOnChangeText = debounce((text) => {
 		const isTextEmpty = text.length === 0;
-		// this.setShowSend(!isTextEmpty);
-		this.handleTyping(!isTextEmpty);
-
-		if (!sharing) {
-			// matches if their is text that stats with '/' and group the command and params so we can use it "/command params"
-			const slashCommand = text.match(/^\/([a-z0-9._-]+) (.+)/im);
-			if (slashCommand) {
```
Comment (diegolmello) [inline app/containers/MessageBox/index.js:365]: Some slash commands run while typing, some run when the message is sent.
While typing, we need to evaluate the kind of command and act as expected.
If you try `/giphy` on https://open.rocket.chat, you'll see that you can play with gif previews while sending a message.
Open the app and type: `/giphy dog`

**Review by diegolmello — CHANGES_REQUESTED** [review]
@sumukhah that's a huge improvement on readability!
Left a few comments.
I can review it again as soon as you address them.
Thank you.

**Comment by sumukhah** [comment]
@diegolmello thank you.. I have made some corrections to the slash command logic.
Please review

**Comment by sumukhah** [comment]
@diegolmello 
made some changes to the slash command logic. Please review
Thank you

**Review by diegolmello — CHANGES_REQUESTED** [review]
@sumukhah Added a few commits to make it return data right after the symbol is recognised and fixed tests.
Just one thing left: since you're not tracking the cursor, a mention in the middle of a sentence doesn't work.

https://user-images.githubusercontent.com/804994/113603669-964ee100-961a-11eb-80ad-2ee2841b292b.mp4

**Comment by sumukhah** [comment]
@diegolmello, Yes, I didn't notice the purpose of that !! 
I have changed some code in this commit. This would fix the issue with the cursor point.
Thank you 👍 

Screenshot 2021-04-06 at 12 12 38 AM

**Comment by sumukhah** [comment]
🎉 I guess this would also fix #1580 now?

**Comment by sumukhah** [comment]
> @sumukhah Added a few commits to make it return data right after the symbol is recognised and fixed tests.

🎉 I guess this would also fix #1580 now?

**Review by diegolmello — APPROVED** [review]
LGTM 🎉

**Comment by ankar84** [comment]
I'm testing 4.16.0 beta and can tell for sure that this PR fixes my issue https://github.com/RocketChat/Rocket.Chat.ReactNative/issues/2529 and probably that issue https://github.com/RocketChat/Rocket.Chat.ReactNative/issues/1580

@sumukhah great job!

**Pull request merged**